### PR TITLE
Core changes for "tctl sso test" command (SAML only).

### PR DESCRIPTION
**Pull request by Tener**
These are the changes required for "tctl sso test" command. For now only implemented for SAML.

See: #9270 for original issue.

Actual command: https://github.com/gravitational/teleport.e/pull/398

Changes to trace library: https://github.com/gravitational/trace/pull/74

### CODE REVIEW

File: lib/auth/clt.go [hunk 1]
```diff
@@ -1176,17 +1176,29 @@ func (c *Client) CreateSAMLAuthRequest(req services.SAMLAuthRequest) (*services.
...
 // ValidateSAMLResponse validates response returned by SAML identity provider
-func (c *Client) ValidateSAMLResponse(re string) (*SAMLAuthResponse, error) {
+func (c *Client) ValidateSAMLResponse(re string) (*SAMLAuthResponse, *SsoDiagnosticInfo, error) {
 	out, err := c.PostJSON(c.Endpoint("saml", "requests", "validate"), validateSAMLResponseReq{
 		Response: re,
 	})
+
+	// try to recover diagnostic info from proxy error
+	var di *SsoDiagnosticInfo
+	di = nil
```
Comment (smallinsky) [inline lib/auth/clt.go:1186]: Di is already nil you don't need to reset it again.

File: lib/auth/clt.go [hunk 2]
```diff
@@ -1176,17 +1176,29 @@ func (c *Client) CreateSAMLAuthRequest(req services.SAMLAuthRequest) (*services.
...
+
+	var ssoDi SsoDiagnosticInfo
+	if trace.UnwrapProxyField(err, "sso-diag-info", &ssoDi) {
```
Comment (smallinsky) [inline lib/auth/clt.go:1189]: missing error check.

File: lib/auth/saml.go [hunk 3]
```diff
@@ -308,80 +346,159 @@ type SAMLAuthResponse struct {
...
+		Success:     false,
+		DiagMap:     map[string]interface{}{},
+		SSOTestFlow: false,
+	}
+
 	requestID, err := parseSAMLInResponseTo(samlResponse)
 	if err != nil {
-		return nil, trace.Wrap(err)
+		di.SetProblem("Failed to parse SAML response", err)
+		return nil, di, trace.Wrap(err)
```
Comment (smallinsky) [inline lib/auth/saml.go:459]: Does the di needs to be returned with err != nil 

What about decoupling error message and successful fileds from `SsoDiagnosticInfo` 
and proagating error to the http caller by  `return nil, trace.Wrap(err).WithField(SSODiagnosticErrors, di)`

File: lib/auth/apiserver.go [hunk 4]
```diff
@@ -1586,23 +1586,29 @@ type samlAuthRawResponse struct {
...
 		return nil, trace.Wrap(err)
 	}
-	response, err := auth.ValidateSAMLResponse(req.Response)
+
+	// we attach the diagnostic information either to the error or response.
+	// this is an unfortunate redundancy caused by limitation of our HTTP API
+	response, di, err := auth.ValidateSAMLResponse(req.Response)
 	if err != nil {
-		return nil, trace.Wrap(err)
+		return nil, trace.Wrap(err).AddField("sso-diag-info", di)
```
Comment (smallinsky) [inline lib/auth/apiserver.go:1603]: Can `di` field can be propagated by the `auth.ValidateSAMLResponse` handler. So the `auth.ValidateSAMLResponse` will return only `response and err` where err will have preset diagnostic ErrorField

File: lib/auth/clt.go (hunk 2 above)
Comment (Tener) [reply lib/auth/clt.go:1189]: UnwrapProxyField returns boolean, not an error?

**Comment by xacrimon** [comment]
If this involved interacting with any SAML response data from a third party IdP like Azure AD, please make sure that this also works when Azure AD Encrypted SAML is enabled.

**Comment by Tener** [comment]
> If this involved interacting with any SAML response data from a third party IdP like Azure AD (which it seems to do), please make sure that this also works when Azure AD Encrypted SAML is enabled. That part of the standard sometimes behaves wierdly and usually needs to be manually tested when making SAML changes.

Thanks for the tip, I'll make sure to go through that flow.

**Comment by Tener** [comment]
Switching to another PR: https://github.com/gravitational/teleport/pull/11508

**Pull request closed without merging**